typescript/schedules.md

```diff
@@ -0,0 +1,1157 @@
+  SKIP,
+
+  /**
+   * Start another Action as soon as the current Action completes, but only buffer one Action in this way. If another
+   * Action is supposed to start, but one Action is running and one is already buffered, then only the buffered one will
+   * be started after the running Action finishes.
+   */
+  BUFFER_ONE,
+
+  /**
+   * Allows an unlimited number of Actions to buffer. They are started sequentially.
+   */
+  BUFFER_ALL,
+
+  /**
+   * Cancels the running Action, and then starts the new Action once the cancelled one completes.
+   */
+  CANCEL_OTHER,
+
+  /**
+   * Terminate the running Action and start the new Action immediately.
+   */
+  TERMINATE_OTHER,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   */
+  ALLOW_ALL,
+}
+
+checkExtendsWithoutPrefix<
+  temporal.api.enums.v1.schedule.ScheduleOverlapPolicy,
+  ScheduleOverlapPolicy,
+  'SCHEDULE_OVERLAP_POLICY_'
+>();
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  start: Date;
+  end: Date;
+
+  /**
+   * Override the Overlap Policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * The range of values depends on which, if any, fields are set:
+ *
+ * ```
+ * {} -> all values
+ * {start} -> start
+ * {start, end} -> every value from start to end (implies step = 1)
+ * {start, step} -> from start, by step (implies end is max for this field)
+ * {start, end, step} -> from start to end, by step
+ * {step} -> all values, by step (implies start and end are full range for this field)
+ * {end} and {end, step} are not allowed
+ * TODO is there a way to express not allowed ^ in type?
+ * ```
+ *
+ * For example:
+ *
+ * ```
+ * {start: 2, end: 10, step: 3} -> 2, 5, 8
+ * ```
+ */
+export interface Range<Unit> {
+  start?: Unit;
+  end?: Unit;
+
+  /**
+   * The step to take between each value.
+   *
+   * @default 1
+   */
+  step?: PositiveInteger;
+}
+
+export type NumberRange = Range<number> | number
+export type NumberSpec = NumberRange | NumberRange[] | string;
+export type NumberSpecDescription = Range<number>[];
+
+export type Month = 'JANUARY' | 'FEBRUARY' | 'MARCH' | 'APRIL' | 'MAY' | 'JUNE' | 'JULY' | 'AUGUST' | 'SEPTEMBER' | 'OCTOBER' | 'NOVEMBER' | 'DECEMBER';
+export type MonthRange = Range<Month> | Month;
+export type MonthSpec = MonthRange | MonthRange[] | string;
+export type MonthSpecDescription = Range<Month>[];
+
+export type Day = 'SUNDAY' | 'MONDAY' | 'TUESDAY' | 'WEDNESDAY' | 'THURSDAY' | 'FRIDAY' | 'SATURDAY';
+export type DayRange = Range<Day> | Day;
+export type DaySpec = DayRange | DayRange[] | string;
+export type DaySpecDescription = Range<Day>[];
+
+/**
+ * An event specification relative to the calendar, similar to a traditional cron specification.
+ *
+ * A second in time matches if all fields match. This includes `dayOfMonth` and `dayOfWeek`.
+ */
+export interface CalendarSpec {
```

Comment:
Does this correspond to the proto `CalendarSpec` or `StructuredCalendarSpec`? I assume the latter? So if a user does
```
{dayOfMonth: "1,15", hour: {start: 12, step: 3}}
```
do you convert the string to ranges on the client side? And are you going to fill in the defaults on the client side too?

Reply:
Structured. Only string accepted by SDK is `'*'`—will update type to:

`export type NumberSpec = NumberRange | NumberRange[] | '*';`

Not planning on filling in defaults—is there a reason to?

Reply:
Also should there be a constant for `'*'`?

Reply:
This will fail compilation due to "1,15":

```
{dayOfMonth: "1,15", hour: {start: 12, step: 3}}
```

And `{hour: [{start: 12, step: 3}]}` will send `{hour: [{start: 12, step: 3}]}` to the server, not filling in defaults. 

> Also should there be a constant for '*'?

Roey and I prefer string literals over constants and enums from a convenience perspective (don't have to import something, clearer type hints)

Reply:
ok, then the docs should say end is required if you want a range of more than one thing. without an explicit end (meaning you'll send end == 0) that means a single value.

it feels slightly weird to have a magic string value without a constant. but I've never written a line of TS in my life, don't listen to me :)

(to be clear and spoil the joke, I've written thousands of lines of JS, just not TS)